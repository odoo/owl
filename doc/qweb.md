# ðŸ¦‰ QWeb ðŸ¦‰

## Content

- [Overview](#overview)
- [Directives](#directives)
- [QWeb Engine](#qweb-engine)
- [Reference](#reference)
  - [White Spaces](#white-spaces)
  - [Root Nodes](#root-nodes)
  - [Expression Evaluation](#expression-evaluation)
  - [Static html Nodes](#static-html-nodes)
  - [Outputting Data](#outputting-data)
  - [Setting Variables](#setting-variables)
  - [Conditionals](#conditionals)
  - [Dynamic Attributes](#dynamic-attributes)
  - [Loops](#loops)
  - [Rendering Sub Templates](#rendering-sub-templates)
  - [Debugging](#debugging)

## Overview

[QWeb](https://www.odoo.com/documentation/13.0/reference/qweb.html) is the primary templating engine used by Odoo. It is based on the XML format, and used
mostly to generate HTML. In OWL, QWeb templates are compiled into functions that
generate a virtual dom representation of the HTML.

Template directives are specified as XML attributes prefixed with `t-`, for instance `t-if` for conditionals, with elements and other attributes being rendered directly.

To avoid element rendering, a placeholder element `<t>` is also available, which executes its directive but doesnâ€™t generate any output in and of itself.

```xml
<div>
    <span t-if="somecondition">Some string</span>
    <ul t-else="1">
        <li t-foreach="messages" t-as="message">
            <t t-esc="message"/>
        </li>
    </ul>
</div>
```

The QWeb class in the OWL project is an implementation of that specification
with a few interesting points:

- it compiles templates into functions that output a virtual DOM instead of a
  string. This is necessary for the component system.
- it has a few extra directives: `t-component`, `t-on`, ...

## Directives

We present here a list of all standard QWeb directives:

| Name                           | Description                                                  |
| ------------------------------ | ------------------------------------------------------------ |
| `t-esc`                        | [Outputting safely a value](#outputting-data)                |
| `t-raw`                        | [Outputting value, without escaping](#outputting-data)       |
| `t-set`, `t-value`             | [Setting variables](#setting-variables)                      |
| `t-if`, `t-elif`, `t-else`,    | [conditionally rendering](#conditionals)                     |
| `t-foreach`, `t-as`            | [Loops](#loops)                                              |
| `t-att`, `t-attf-*`, `t-att-*` | [Dynamic attributes](#dynamic-attributes)                    |
| `t-call`                       | [Rendering sub templates](#rendering-sub-templates)          |
| `t-debug`, `t-log`             | [Debugging](#debugging)                                      |
| `t-name`                       | [Defining a template (not really a directive)](#qweb-engine) |

The component system in Owl requires additional directives, to express various
needs. Here is a list of all Owl specific directives:

| Name                                                   | Description                                                                         |
| ------------------------------------------------------ | ----------------------------------------------------------------------------------- |
| `t-component`, `t-props`, `t-keepalive`, `t-asyncroot` | [Defining a sub component](component.md#composition)                                |
| `t-ref`                                                | [Setting a reference to a dom node or a sub component](component.md#references)     |
| `t-key`                                                | [Defining a key (to help virtual dom reconciliation)](component.md#t-key-directive) |
| `t-on-*`                                               | [Event handling](component.md#event-handling)                                       |
| `t-transition`                                         | [Defining an animation](animations.md#css-transitions)                              |
| `t-slot`                                               | [Rendering a slot](component.md#slots)                                              |
| `t-model`                                              | [Form input bindings](component.md#form-input-bindings)                             |

## QWeb Engine

This section is about the javascript code that implements the `QWeb` specification.
Owl exports a `QWeb` class in `owl.QWeb`. To use it, it just needs to be
instantiated:

```js
const qweb = new owl.QWeb();
```

Its API is quite simple:

- **`constructor(data)`**: constructor. Takes an optional string to add initial
  templates (see `addTemplates` for more information on format of the string).

  ```js
  const qweb = new owl.QWeb(TEMPLATES);
  ```

- **`addTemplate(name, xmlStr, allowDuplicate)`**: add a specific template.

  ```js
  qweb.addTemplate("mytemplate", "<div>hello</div>");
  ```

  If the optional `allowDuplicate` is set to `true`, then `QWeb` will simply
  ignore templates added for a second time. Otherwise, `QWeb` will crash.

- **`addTemplates(xmlStr)`**: add a list of templates (identified by `t-name`
  attribute).

  ```js
  const TEMPLATES = `
    <templates>
      <div t-name="App" class="main">main</div>
      <div t-name="OtherComponent">other component</div>
    </templates>`;
  qweb.addTemplates(TEMPLATES);
  ```

- **`render(name, context, extra)`**: renders a template. This returns a `vnode`,
  which is a virtual representation of the DOM (see [vdom doc](vdom.md)).

  ```js
  const vnode = qweb.render("App", component);
  ```

- **`renderToString(name, context)`**: renders a template, but returns an html
  string.

  ```js
  const str = qweb.renderToString("someTemplate", somecontext);
  ```

- **`registerTemplate(name, template)`**: static function to register a global
  QWeb template. This is useful for commonly used components accross the
  application, and for making a template available to an application without
  having a reference to the actual QWeb instance.

  ```js
  QWeb.registerTemplate("mytemplate", `<div>some template</div>`);
  ```

- **`registerComponent(name, Component)`**: static function to register an OWL Component
  to QWeb's global registry. Globally registered Components can be used in
  templates (see the `t-component` directive). This is useful for commonly used
  components accross the application.

  ```js
  class Dialog extends owl.Component { ... }
  QWeb.registerComponent("Dialog", Dialog);

  ...

  class ParentComponent extends owl.Component { ... }
  qweb.addTemplate("ParentComponent", "<div><Dialog/></div>");
  ```

In some way, a `QWeb` instance is the core of an Owl application. It is the only
mandatory element of an [environment](component.md#environment). As such, it
has an extra responsibility: it can act as an event bus for internal communication
between Owl classes. This is the reason why `QWeb` actually extends [EventBus](event_bus.md).

## Reference

We define in this section the specification of how `QWeb` templates should be
rendered. Note that we only document here the standard QWeb specification. Owl
specific extensions are documented in various other parts of the documentation.

### White Spaces

White spaces in a template are handled in a special way:

- consecutive whitespaces are always condensed to a single whitespace
- if a whitespace-only text node contains a linebreak, it is ignored
- the previous rules do not apply if we are in a `<pre>` tag

### Root Nodes

For many reasons, Owl QWeb templates should have a single root node. More
precisely, the result of a template rendering should have a single root node:

```xml
<!â€“â€“ not ok: two root nodes â€“â€“>
<t>
    <div>foo</div>
    <div>bar</div>
</t>

<!â€“â€“ ok: result has one single root node â€“â€“>
<t>
    <div t-if="someCondition">foo</div>
    <span t-else="1">bar</span>
</t>
```

Extra root nodes will actually be ignored (even though they will be rendered
in memory).

Note: this does not apply to subtemplates (see the `t-call` directive). In that
case, they will be inlined in the main template, and can actually have many
root nodes.

### Expression Evaluation

QWeb expressions are strings that will be processed at compile time. Each variable in
the javascript expression will be replaced by a lookup in the context (so, the
component). For example, `a + b.c(d)` will be converted into:

```js
context["a"] + context["b"].c(context["d"]);
```

It is useful to explain the various rules that apply on these expressions:

1. it should be a simple expression which returns a value. It cannot be a statement.

   ```xml
   <div><p t-if="1 + 2 === 3">ok</p></div>
   ```

   is valid, but the following is not valid:

   ```xml
   <div><p t-if="console.log(1)">NOT valid</p></div>
   ```

2. it can use anything in the rendering context (typically, the component):

   ```xml
   <p t-if="user.birthday === today()">Happy bithday!</p>
   ```

   is valid, and will read the `user` object from the context, and call the
   `today` function.

3. it can use a few special operators to avoid using symbols such as `<`, `>`,
   `&` or `|`. This is useful to make sure that we still write valid XML.

   | Word  | will be replaced by |
   | ----- | ------------------- |
   | `and` | `&&`                |
   | `or`  | `\|\|`              |
   | `gt`  | `>`                 |
   | `gte` | `>=`                |
   | `lt`  | `<`                 |
   | `lte` | `<=`                |

   So, one can write this:

   ```xml
   <div><p t-if="10 + 2 gt 5">ok</p></div>
   ```

### Static Html Nodes

Normal, regular html nodes are rendered into themselves:

```xml
  <div>hello</div> <!â€“â€“ rendered as itself â€“â€“>
```

### Outputting Data

The `t-esc` directive is necessary whenever you want to add a dynamic text
expression in a template. The text is escaped to avoid security issues.

```xml
<p><t t-esc="value"/></p>
```

rendered with the value `value` set to `42` in the rendering context yields:

```html
<p>42</p>
```

The `t-raw` directive is almost the same as `t-esc`, but without the escaping.
This is mostly useful to inject a raw html string somewhere. Obviously, this
is unsafe to do in general, and should only be used for strings known to be safe.

```xml
<p><t t-raw="value"/></p>
```

rendered with the value `value` set to `<span>foo</span>` in the rendering context yields:

```html
<p><span>foo</span></p>
```

Note that since the content of the expression is not known beforehand, the `t-raw`
directive has to parse the html (and convert it to a virtual dom structure) for
each rendering. So, it will be much slower than a regular template. It is
therefore advised to limit the use of `t-raw` whenever possible.

### Setting Variables

QWeb allows creating variables from within the template, to memoize a computation (to use it multiple times), give a piece of data a clearer name, ...

This is done via the `t-set` directive, which takes the name of the variable to create. The value to set can be provided in two ways:

1. a `t-value` attribute containing an expression, and the result of its
   evaluation will be set:

   ```xml
   <t t-set="foo" t-value="2 + 1"/>
   <t t-esc="foo"/>
   ```

   will print `3`. Note that the evaluation is done at rendering time, not at
   compilte time.

2. if there is no `t-value` attribute, the nodeâ€™s body is saved and its value is
   set as the variableâ€™s value:

   ```xml
   <t t-set="foo">
       <li>ok</li>
   </t>
   <t t-esc="foo"/>
   ```

   will generate `&lt;li&gt;ok&lt;/li&gt;` (the content is escaped as we used the `t-esc` directive)

The `t-set` directive acts like a regular variable in most programming language.
It is lexically scoped (inner nodes are sub scopes), can be shadowed, ...

### Conditionals

The `t-if` directive is useful to conditionally render something. It evaluates
the expression given as attribute value, and then acts accordingly.

```xml
<div>
    <t t-if="condition">
        <p>ok</p>
    </t>
</div>
```

The element is rendered if the condition (evaluated with the current rendering
context) is true:

```xml
<div>
    <p>ok</p>
</div>
```

but if the condition is false it is removed from the result:

```xml
<div>
</div>
```

The conditional rendering applies to the bearer of the directive, which does not
have to be `<t>`:

```xml
<div>
    <p t-if="condition">ok</p>
</div>
```

will give the same results as the previous example.

Extra conditional branching directives `t-elif` and `t-else` are also available:

```xml
<div>
    <p t-if="user.birthday == today()">Happy bithday!</p>
    <p t-elif="user.login == 'root'">Welcome master!</p>
    <p t-else="">Welcome!</p>
</div>
```

### Dynamic Attributes

One can use the `t-att-` directive to add dynamic attributes. Its main use is to
evaluate an expression (at rendering time) and bind an attribute to its result:

For example, if we have `id` set to 32 in the rendering context,

```xml
<div t-att-data-action-id="id"/> <!-- result: <div data-action-id="32"></div> -->
```

If an expression evaluates to a falsy value, it will not be set at all:

```xml
<div t-att-foo="false"/>  <!-- result: <div></div> -->
```

There is another way to format a string attribute: the `t-attf-` directive. With
it, you get string interpolation:

```xml
<div t-attf-foo="a {{value1}} is {{value2}} of {{value3}} ]"/>
<!-- result if values are set to 1,2 and 3: <div foo="a 0 is 1 of 2 ]"></div> -->
```

### Loops

QWeb has an iteration directive `t-foreach` which take an expression returning the
collection to iterate on, and a second parameter `t-as` providing the name to use
for the current item of the iteration:

```xml
<t t-foreach="[1, 2, 3]" t-as="i">
    <p><t t-esc="i"/></p>
</t>
```

will be rendered as:

```xml
<p>1</p>
<p>2</p>
<p>3</p>
```

Like conditions, `t-foreach` applies to the element bearing the directiveâ€™s attribute, and

```xml
<p t-foreach="[1, 2, 3]" t-as="i">
    <t t-esc="i"/>
</p>
```

is equivalent to the previous example.

`t-foreach` can iterate on an array (the current item will be the current value)
or an object (the current item will be the current key).

In addition to the name passed via t-as, `t-foreach` provides a few other
variables for various data points (note: `$as` will be replaced by the name
passed to `t-as`):

- `$as_value`: the current iteration value, identical to `$as` for lists and
  integers, but for objects, it provides the value (where `$as` provides the key)
- `$as_index`: the current iteration index (the first item of the iteration has index 0)
- `$as_first`: whether the current item is the first of the iteration
  (equivalent to `$as_index == 0`)
- `$as_last`: whether the current item is the last of the iteration
  (equivalent to `$as_index + 1 == $as_size`), requires the iterateeâ€™s size be
  available

These extra variables provided and all new variables created into the `t-foreach`
are only available in the scope of the `t-foreach`. If the variable exists outside
the context of the `t-foreach`, the value is copied at the end of the foreach
into the global context.

```xml
<t t-set="existing_variable" t-value="False"/>
<!-- existing_variable now False -->

<p t-foreach="Array(3)" t-as="i">
    <t t-set="existing_variable" t-value="True"/>
    <t t-set="new_variable" t-value="True"/>
    <!-- existing_variable and new_variable now True -->
</p>

<!-- existing_variable always True -->
<!-- new_variable undefined -->
```

### Rendering Sub Templates

QWeb templates can be used for top level rendering, but they can also be used
from within another template (to avoid duplication or give names to parts of
templates), using the `t-call` directive:

```xml
<div t-name="other-template">
    <p><t t-value="var"/></p>
</div>

<div t-name="main-template">
    <t t-set="var" t-value="owl"/>
    <t t-call="other-template"/>
</div>
```

will be rendered as `<div><p>owl</p></div>`. This example shows that the sub
template is rendered with the execution context of the parent. The sub template
is actually inlined in the main template, but in a sub scope: variables defined
in the sub template do not escape.

Sometimes, one might want to pass information to the sub template. In that case,
the content of the body of the `t-call` directive is available as a special
magic variable `0`:

```xml
<t t-name="other-template">
    This template was called with content:
    <t t-raw="0"/>
</t>

<div t-name="main-template">
    <t t-call="other-template">
        <em>content</em>
    </t>
</div>
```

will result in :

```xml
<div>
    This template was called with content:
    <em>content</em>
</div>
```

### Debugging

The javascript QWeb implementation provides two useful debugging directives:

`t-debug` adds a debugger statement during template rendering:

```xml
<t t-if="a_test">
    <t t-debug="">
</t>
```

will stop execution if the browser dev tools are open.

`t-log` takes an expression parameter, evaluates the expression during rendering and logs its result with console.log:

```xml
<t t-set="foo" t-value="42"/>
<t t-log="foo"/>
```

will print 42 to the console.
