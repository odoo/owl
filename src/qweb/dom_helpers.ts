export const enum DomType {
  Text,
  Comment,
  Node,
}

export interface DomText {
  type: DomType.Text;
  value: string;
}

export interface DomComment {
  type: DomType.Comment;
  value: string;
}
export interface DomNode {
  type: DomType.Node;
  tag: string;
  attrs: { [key: string]: string };
  content: Dom[];
}

export type Dom = DomText | DomComment | DomNode;

function escape(str: string): string {
  const p = document.createElement("p");
  p.textContent = str;
  return p.innerHTML;
}

export function domToString(dom: Dom): string {
  switch (dom.type) {
    case DomType.Text:
      return escape(dom.value);
    case DomType.Comment:
      return `<!--${dom.value}-->`;
    case DomType.Node:
      const content = dom.content.map(domToString).join("");
      const attrs: string[] = [];
      for (let [key, value] of Object.entries(dom.attrs)) {
        if (!(key === "class" && value === "")) {
          attrs.push(`${key}="${escape(value)}"`);
        }
      }
      if (content) {
        return `<${dom.tag}${attrs.length ? " " + attrs.join(" ") : ""}>${content}</${dom.tag}>`;
      } else {
        return `<${dom.tag}${attrs.length ? " " + attrs.join(" ") : ""}/>`;
      }
  }
}

export function isProp(tag: string, key: string): boolean {
  switch (tag) {
    case "input":
      return (
        key === "checked" ||
        key === "indeterminate" ||
        key === "value" ||
        key === "readonly" ||
        key === "disabled"
      );
    case "option":
      return key === "selected" || key === "disabled";
    case "textarea":
      return key === "readonly" || key === "disabled";
      break;
    case "button":
    case "select":
    case "optgroup":
      return key === "disabled";
  }
  return false;
}
