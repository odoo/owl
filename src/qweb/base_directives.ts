import { CompilationContext } from "./compilation_context";
import { QWeb } from "./qweb";
import { htmlToVDOM } from "../vdom/html_to_vdom";
import { QWebVar } from "./expression_parser";

/**
 * Owl QWeb Directives
 *
 * This file contains the implementation of most standard QWeb directives:
 * - t-esc
 * - t-raw
 * - t-set/t-value
 * - t-if/t-elif/t-else
 * - t-call
 * - t-foreach/t-as
 * - t-debug
 * - t-log
 */

//------------------------------------------------------------------------------
// t-esc and t-raw
//------------------------------------------------------------------------------
QWeb.utils.htmlToVDOM = htmlToVDOM;

function compileValueNode(value: any, node: Element, qweb: QWeb, ctx: CompilationContext) {
  ctx.rootContext.shouldDefineScope = true;
  if (value === "0") {
    if (ctx.parentNode) {
      // the 'zero' magical symbol is where we can find the result of the rendering
      // of  the body of the t-call.
      ctx.rootContext.shouldDefineUtils = true;
      const zeroArgs = ctx.escaping
        ? `{text: utils.vDomToString(scope[utils.zero])}`
        : `...scope[utils.zero]`;
      ctx.addLine(`c${ctx.parentNode}.push(${zeroArgs});`);
    }
    return;
  }

  let exprID: string;
  if (typeof value === "string") {
    exprID = `_${ctx.generateID()}`;
    ctx.addLine(`let ${exprID} = ${ctx.formatExpression(value)};`);
  } else {
    exprID = `scope.${value.id}`;
  }
  ctx.addIf(`${exprID} != null`);

  if (ctx.escaping) {
    let protectID;
    if (value.hasBody) {
      protectID = ctx.startProtectScope();
      ctx.addLine(
        `${exprID} = ${exprID} instanceof utils.VDomArray ? utils.vDomToString(${exprID}) : ${exprID};`
      );
    }
    if (ctx.parentTextNode) {
      ctx.addLine(`vn${ctx.parentTextNode}.text += ${exprID};`);
    } else if (ctx.parentNode) {
      ctx.addLine(`c${ctx.parentNode}.push({text: ${exprID}});`);
    } else {
      let nodeID = ctx.generateID();
      ctx.rootContext.rootNode = nodeID;
      ctx.rootContext.parentTextNode = nodeID;
      ctx.addLine(`let vn${nodeID} = {text: ${exprID}};`);
      if (ctx.rootContext.shouldDefineResult) {
        ctx.addLine(`result = vn${nodeID}`);
      }
    }
    if (value.hasBody) {
      ctx.stopProtectScope(protectID);
    }
  } else {
    ctx.rootContext.shouldDefineUtils = true;
    if (value.hasBody) {
      ctx.addLine(
        `const vnodeArray = ${exprID} instanceof utils.VDomArray ? ${exprID} : utils.htmlToVDOM(${exprID});`
      );
      ctx.addLine(`c${ctx.parentNode}.push(...vnodeArray);`);
    } else {
      ctx.addLine(`c${ctx.parentNode}.push(...utils.htmlToVDOM(${exprID}));`);
    }
  }
  if (node.childNodes.length) {
    ctx.addElse();
    qweb._compileChildren(node, ctx);
  }

  ctx.closeIf();
}

QWeb.addDirective({
  name: "esc",
  priority: 70,
  atNodeEncounter({ node, qweb, ctx }): boolean {
    let value = ctx.getValue(node.getAttribute("t-esc")!);
    compileValueNode(value, node, qweb, ctx.subContext("escaping", true));
    return true;
  },
});

QWeb.addDirective({
  name: "raw",
  priority: 80,
  atNodeEncounter({ node, qweb, ctx }): boolean {
    let value = ctx.getValue(node.getAttribute("t-raw")!);
    compileValueNode(value, node, qweb, ctx);
    return true;
  },
});

//------------------------------------------------------------------------------
// t-set
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "set",
  extraNames: ["value"],
  priority: 60,
  atNodeEncounter({ node, qweb, ctx }): boolean {
    ctx.rootContext.shouldDefineScope = true;
    const variable = node.getAttribute("t-set")!;
    let value = node.getAttribute("t-value")!;
    ctx.variables[variable] = ctx.variables[variable] || ({} as QWebVar);
    let qwebvar = ctx.variables[variable];
    const hasBody = node.hasChildNodes();

    qwebvar.id = variable;
    qwebvar.expr = `scope.${variable}`;
    if (value) {
      const formattedValue = ctx.formatExpression(value);
      let scopeExpr = `scope`;
      if (ctx.protectedScopeNumber) {
        ctx.rootContext.shouldDefineUtils = true;
        scopeExpr = `utils.getScope(scope, '${variable}')`;
      }
      ctx.addLine(`${scopeExpr}.${variable} = ${formattedValue};`);
      qwebvar.value = formattedValue;
    }

    if (hasBody) {
      ctx.rootContext.shouldDefineUtils = true;
      if (value) {
        ctx.addIf(`!(${qwebvar.expr})`);
      }
      const tempParentNodeID = ctx.generateID();
      const _parentNode = ctx.parentNode;
      ctx.parentNode = tempParentNodeID;

      ctx.addLine(`let c${tempParentNodeID} = new utils.VDomArray();`);
      const nodeCopy = node.cloneNode(true) as Element;
      for (let attr of ["t-set", "t-value", "t-if", "t-else", "t-elif"]) {
        nodeCopy.removeAttribute(attr);
      }
      qweb._compileNode(nodeCopy, ctx);

      ctx.addLine(`${qwebvar.expr} = c${tempParentNodeID}`);
      qwebvar.value = `c${tempParentNodeID}`;
      qwebvar.hasBody = true;

      ctx.parentNode = _parentNode;
      if (value) {
        ctx.closeIf();
      }
    }
    return true;
  },
});

//------------------------------------------------------------------------------
// t-if, t-elif, t-else
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "if",
  priority: 20,
  atNodeEncounter({ node, ctx }): boolean {
    let cond = ctx.getValue(node.getAttribute("t-if")!);
    ctx.addIf(typeof cond === "string" ? ctx.formatExpression(cond) : `scope.${cond.id!}`);
    return false;
  },
  finalize({ ctx }) {
    ctx.closeIf();
  },
});

QWeb.addDirective({
  name: "elif",
  priority: 30,
  atNodeEncounter({ node, ctx }): boolean {
    let cond = ctx.getValue(node.getAttribute("t-elif")!);
    ctx.addLine(
      `else if (${typeof cond === "string" ? ctx.formatExpression(cond) : `scope.${cond.id}`}) {`
    );
    ctx.indent();
    return false;
  },
  finalize({ ctx }) {
    ctx.closeIf();
  },
});

QWeb.addDirective({
  name: "else",
  priority: 40,
  atNodeEncounter({ ctx }): boolean {
    ctx.addLine(`else {`);
    ctx.indent();
    return false;
  },
  finalize({ ctx }) {
    ctx.closeIf();
  },
});

//------------------------------------------------------------------------------
// t-call
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "call",
  priority: 50,
  atNodeEncounter({ node, qweb, ctx }): boolean {
    // Step 1: sanity checks
    // ------------------------------------------------
    ctx.rootContext.shouldDefineScope = true;
    ctx.rootContext.shouldDefineUtils = true;
    if (node.nodeName !== "t") {
      throw new Error("Invalid tag for t-call directive (should be 't')");
    }
    const subTemplate = node.getAttribute("t-call")!;
    const nodeTemplate = qweb.templates[subTemplate];
    if (!nodeTemplate) {
      throw new Error(`Cannot find template "${subTemplate}" (t-call)`);
    }

    // Step 2: compile target template in sub templates
    // ------------------------------------------------
    if (!qweb.subTemplates[subTemplate]) {
      qweb.subTemplates[subTemplate] = true;
      const subTemplateFn = qweb._compile(subTemplate, nodeTemplate.elem, ctx, true);
      qweb.subTemplates[subTemplate] = subTemplateFn;
    }

    // Step 3: compile t-call body if necessary
    // ------------------------------------------------
    let hasBody = node.hasChildNodes();
    let protectID;
    if (hasBody) {
      // we add a sub scope to protect the ambient scope
      ctx.addLine(`{`);
      ctx.indent();
      protectID = ctx.startProtectScope();
      const nodeCopy = node.cloneNode(true) as Element;
      for (let attr of ["t-if", "t-else", "t-elif", "t-call"]) {
        nodeCopy.removeAttribute(attr);
      }
      const parentNode = ctx.parentNode;
      ctx.parentNode = "__0";
      // this local scope is intended to trap c__0
      ctx.addLine(`{`);
      ctx.indent();
      ctx.addLine("let c__0 = [];");
      qweb._compileNode(nodeCopy, ctx);
      ctx.rootContext.shouldDefineUtils = true;
      ctx.addLine("scope[utils.zero] = c__0;");
      ctx.parentNode = parentNode;
      ctx.dedent();
      ctx.addLine(`}`);
    }

    // Step 4: add the appropriate function call to current component
    // ------------------------------------------------
    const callingScope = hasBody ? "scope" : "Object.assign(Object.create(context), scope)";
    const parentComponent = `utils.getComponent(context)`;
    const key = ctx.generateTemplateKey();
    const parentNode = ctx.parentNode ? `c${ctx.parentNode}` : "result";
    const extra = `Object.assign({}, extra, {parentNode: ${parentNode}, parent: ${parentComponent}, key: ${key}})`;
    if (ctx.parentNode) {
      ctx.addLine(`this.subTemplates['${subTemplate}'].call(this, ${callingScope}, ${extra});`);
    } else {
      // this is a t-call with no parentnode, we need to extract the result
      ctx.rootContext.shouldDefineResult = true;
      ctx.addLine(`result = []`);
      ctx.addLine(`this.subTemplates['${subTemplate}'].call(this, ${callingScope}, ${extra});`);
      ctx.addLine(`result = result[0]`);
    }

    // Step 5: restore previous scope
    // ------------------------------------------------
    if (hasBody) {
      ctx.stopProtectScope(protectID);
      ctx.dedent();
      ctx.addLine(`}`);
    }

    return true;
  },
});

//------------------------------------------------------------------------------
// t-foreach
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "foreach",
  extraNames: ["as"],
  priority: 10,
  atNodeEncounter({ node, qweb, ctx }): boolean {
    ctx.rootContext.shouldDefineScope = true;
    ctx = ctx.subContext("loopNumber", ctx.loopNumber + 1);
    const elems = node.getAttribute("t-foreach")!;
    const name = node.getAttribute("t-as")!;
    let arrayID = ctx.generateID();
    ctx.addLine(`let _${arrayID} = ${ctx.formatExpression(elems)};`);
    ctx.addLine(`if (!_${arrayID}) { throw new Error('QWeb error: Invalid loop expression')}`);
    let keysID = ctx.generateID();
    let valuesID = ctx.generateID();
    ctx.addLine(`let _${keysID} = _${valuesID} = _${arrayID};`);
    ctx.addIf(`!(_${arrayID} instanceof Array)`);
    ctx.addLine(`_${keysID} = Object.keys(_${arrayID});`);
    ctx.addLine(`_${valuesID} = Object.values(_${arrayID});`);
    ctx.closeIf();
    ctx.addLine(`let _length${keysID} = _${keysID}.length;`);
    let varsID = ctx.startProtectScope(true);
    const loopVar = `i${ctx.loopNumber}`;
    ctx.addLine(`for (let ${loopVar} = 0; ${loopVar} < _length${keysID}; ${loopVar}++) {`);
    ctx.indent();

    ctx.addLine(`scope.${name}_first = ${loopVar} === 0`);
    ctx.addLine(`scope.${name}_last = ${loopVar} === _length${keysID} - 1`);
    ctx.addLine(`scope.${name}_index = ${loopVar}`);
    ctx.addLine(`scope.${name} = _${keysID}[${loopVar}]`);
    ctx.addLine(`scope.${name}_value = _${valuesID}[${loopVar}]`);
    const nodeCopy = <Element>node.cloneNode(true);
    let shouldWarn =
      !nodeCopy.hasAttribute("t-key") &&
      node.children.length === 1 &&
      node.children[0].tagName !== "t" &&
      !node.children[0].hasAttribute("t-key");
    if (shouldWarn) {
      console.warn(
        `Directive t-foreach should always be used with a t-key! (in template: '${ctx.templateName}')`
      );
    }
    if (nodeCopy.hasAttribute("t-key")) {
      const expr = ctx.formatExpression(nodeCopy.getAttribute("t-key")!);
      ctx.addLine(`let key${ctx.loopNumber} = ${expr};`);
      nodeCopy.removeAttribute("t-key");
    } else {
      ctx.addLine(`let key${ctx.loopNumber} = i${ctx.loopNumber};`);
    }

    nodeCopy.removeAttribute("t-foreach");
    qweb._compileNode(nodeCopy, ctx);
    ctx.dedent();
    ctx.addLine("}");
    ctx.stopProtectScope(varsID);
    return true;
  },
});

//------------------------------------------------------------------------------
// t-debug
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "debug",
  priority: 1,
  atNodeEncounter({ ctx }) {
    ctx.addLine("debugger;");
  },
});

//------------------------------------------------------------------------------
// t-log
//------------------------------------------------------------------------------
QWeb.addDirective({
  name: "log",
  priority: 1,
  atNodeEncounter({ ctx, value }) {
    const expr = ctx.formatExpression(value);
    ctx.addLine(`console.log(${expr})`);
  },
});
