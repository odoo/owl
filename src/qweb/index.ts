import "./base_directives";
import "./extensions";

export { CompiledTemplate, QWeb } from "./qweb";
