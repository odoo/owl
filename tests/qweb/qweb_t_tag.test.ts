import { QWeb } from "../../src/qweb/index";
import { renderToString } from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

function render(template, context = {}) {
  const qweb = new QWeb();
  qweb.addTemplate("test", template);
  return renderToString(qweb, "test", context);
}

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe("qweb t-tag", () => {
  test("simple usecases", () => {
    expect(render(`<t t-tag="'div'"></t>`)).toBe("<div></div>");
    expect(render(`<t t-tag="tag">text</t>`, { tag: "span" })).toBe("<span>text</span>");
  });

  test("with multiple child nodes", () => {
    const template = `
            <t t-tag="tag">
                pear
                <span>apple</span>
                strawberry
            </t>`;
    expect(render(template, { tag: "div" })).toBe(
      "<div> pear <span>apple</span> strawberry </div>"
    );
  });

  test("with multiple attributes", () => {
    const template = `
            <t t-tag="tag" class="blueberry" taste="raspberry">gooseberry</t>`;
    const expected = `<div taste=\"raspberry\" class=\"blueberry\">gooseberry</div>`;
    expect(render(template, { tag: "div" })).toBe(expected);
  });
});
