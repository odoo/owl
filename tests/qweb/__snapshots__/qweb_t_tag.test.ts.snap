// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qweb t-tag simple usecases 1`] = `
"function anonymous(context, extra
) {
    // Template name: \\"test\\"
    let scope = Object.create(context);
    let result;
    let h = this.h;
    let c1 = [], p1 = {key:1};
    let tag2 = 'div';
    let vn1 = h(tag2, p1, c1);
    result = vn1;
    return result;
}"
`;

exports[`qweb t-tag simple usecases 2`] = `
"function anonymous(context, extra
) {
    // Template name: \\"test\\"
    let scope = Object.create(context);
    let result;
    let h = this.h;
    let c3 = [], p3 = {key:3};
    let tag4 = scope['tag'];
    let vn3 = h(tag4, p3, c3);
    result = vn3;
    c3.push({text: \`text\`});
    return result;
}"
`;

exports[`qweb t-tag with multiple attributes 1`] = `
"function anonymous(context, extra
) {
    // Template name: \\"test\\"
    let scope = Object.create(context);
    let result;
    let h = this.h;
    let _2 = {'blueberry':true};
    let _3 = 'raspberry';
    let c4 = [], p4 = {key:4,attrs:{taste: _3},class:_2};
    let tag5 = scope['tag'];
    let vn4 = h(tag5, p4, c4);
    result = vn4;
    c4.push({text: \`gooseberry\`});
    return result;
}"
`;

exports[`qweb t-tag with multiple child nodes 1`] = `
"function anonymous(context, extra
) {
    // Template name: \\"test\\"
    let scope = Object.create(context);
    let result;
    let h = this.h;
    let c1 = [], p1 = {key:1};
    let tag2 = scope['tag'];
    let vn1 = h(tag2, p1, c1);
    result = vn1;
    c1.push({text: \` pear \`});
    let c3 = [], p3 = {key:3};
    let vn3 = h('span', p3, c3);
    c1.push(vn3);
    c3.push({text: \`apple\`});
    c1.push({text: \` strawberry \`});
    return result;
}"
`;
